trinity/tools/async_process_runner.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+import os
+import signal
+from typing import (
+    AsyncIterable,
+    Awaitable,
+    Callable,
+    Iterable,
+)
+
+
+AwaitableGeneratorProducer = Callable[[], Awaitable[bytes]]
+
+
+class AsyncProcessRunner():
+
+    @classmethod
+    async def create_and_run(cls, cmds: Iterable[str], timeout_sec: int=10) -> 'AsyncProcessRunner':
+        runner = cls()
+        await runner.run(cmds, timeout_sec)
+        return runner
+
+    async def run(self, cmds: Iterable[str], timeout_sec: int=10) -> None:
+        proc = await asyncio.create_subprocess_exec(
+            *cmds,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            # We need this because Trinity spawns multiple processes and we need to take down
+            # the entire group of processes.
+            preexec_fn=os.setsid
+        )
+        self.proc = proc
+        asyncio.ensure_future(self.kill_after_timeout(timeout_sec))
+
+    async def iterate_stdout(self) -> AsyncIterable[str]:
```

Comment:
I wonder if the following three methods were better named `yield_*` instead of `iterate_*` because now that I'm thinking about it, I find the `iterate` prefix rather wrong. A *consumer* would iterate over the produced `AsyncIterable` but that has nothing to do with the method itself which just asynchronously yield stuff. Thoughts?

Reply:
How about just making them properties which return async iterators.  Since the class is `AsyncProcessRunner` I think it's not too big of a stretch for the `runner.stdout` to return something that supports async iteration.

Reply:
I agree. Other options:
- `stdout_lines` <- my fave
- `stdout_buffer`
- `stdout_iterable`